Fortinet Filebeat Module

P1llus opened the pull request:
## What does this PR do?

This PR Introduces the Fortinet filebeat module. Focusing currently on Fortigate Firewall, but should include other Fortinet products as separate PR's later on.

Many thanks to the continuous support from @enotspe and the project https://github.com/enotspe/fortinet-2-elasticsearch helping out with feedback, comments, documentation and much more!

## Why is it important?

Adding more supported products to the filebeat portfolio.

## Checklist


- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Nosetest passes with:
`
INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false TESTING_FILEBEAT_MODULES=fortinet nosetests -v -s tests/system/test_xpack_modules.py
`

## Related issues

- Closes #13245
- Closes elastic/ecs#492

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

P1llus commented on x-pack/filebeat/module/fortinet/firewall/_meta/fields.yml:
```diff
@@ -0,0 +1,2154 @@
+      description: >
+        Some description
+
+    - name: cveid
```
I think the only issue I had here was that since there was no dataset to backup that specific field, I was a bit hesitant to add it directly. Some fortinet modules for example would include an array of CVE's, but ECS do not want that field as an array at this point.

elasticmachine commented:
## :broken_heart: Build Failed 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/6//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/6//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/6//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/6//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_17890.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #17890 updated]
* Start Time: 2020-04-29T12:24:56.044+0000

* Duration: 65 min 21 sec (3920583)


* Commit: abf6f9a036b915cd3a5c3a31ccbf59f51058710c



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 2768  |
| Skipped      | 404 |
| Total        | 3173   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors
  <details><summary>Expand to view the tests failures</summary>
  <p>
    
* **Name**: Build and Test / Filebeat Windows / test_clean_removed – test_registrar.Test
  * Status: FAILED
  * Age: 1
  * Duration: 0.251
  * Error Details: [Errno 2] No such file or directory: 'C:\\Users\\jenkins\\workspace\\Beats_beats-beats-mbp_PR-17890\\src\\github.com\\elastic\\beats\\filebeat\\build\\system-tests\\run\\test_registrar.Test.test_clean_removed\\registry/filebeat/data.json'
    
  </p>
  </details>


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: Mage build unitTest
    * Description: mage  build unitTest
    * Result: FAILURE
    
    * Duration: 10 min 42 sec<
    
    * Start Time: 2020-04-29T13:06:32.719+0000
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-04-29T13:29:47.389Z] Recording test results
[2020-04-29T13:30:01.658Z] Archiving artifacts
[2020-04-29T13:30:02.120Z] java.lang.InterruptedException: no matches found within 10000
[2020-04-29T13:30:02.120Z] 	at hudson.FilePath$ValidateAntFileMask.hasMatch(FilePath.java:2826)
[2020-04-29T13:30:02.120Z] 	at hudson.FilePath$ValidateAntFileMask.invoke(FilePath.java:2705)
[2020-04-29T13:30:02.120Z] 	at hudson.FilePath$ValidateAntFileMask.invoke(FilePath.java:2686)
[2020-04-29T13:30:02.120Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3073)
[2020-04-29T13:30:02.120Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from beats-ci-immutable-ubuntu-1604-1588164438923247670.c.elastic-ci-prod.internal/10.224.2.63:54266
[2020-04-29T13:30:02.120Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1788)
[2020-04-29T13:30:02.120Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
[2020-04-29T13:30:02.120Z] 		at hudson.remoting.Channel.call(Channel.java:998)
[2020-04-29T13:30:02.120Z] 		at hudson.FilePath.act(FilePath.java:1069)
[2020-04-29T13:30:02.120Z] 		at hudson.FilePath.act(FilePath.java:1058)
[2020-04-29T13:30:02.120Z] 		at hudson.FilePath.validateAntFileMask(FilePath.java:2684)
[2020-04-29T13:30:02.120Z] 		at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:265)
[2020-04-29T13:30:02.120Z] 		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
[2020-04-29T13:30:02.120Z] 		at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
[2020-04-29T13:30:02.120Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2020-04-29T13:30:02.120Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2020-04-29T13:30:02.120Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-04-29T13:30:02.120Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-04-29T13:30:02.120Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-04-29T13:30:02.120Z] Caused: hudson.FilePath$TunneledInterruptedException
[2020-04-29T13:30:02.120Z] 	at hudson.FilePath$FileCallableWrapper.call(FilePath.java:3075)
[2020-04-29T13:30:02.120Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
[2020-04-29T13:30:02.120Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2020-04-29T13:30:02.120Z] 	at hudson.remoting.Request$2.run(Request.java:369)
[2020-04-29T13:30:02.120Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-04-29T13:30:02.121Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:93)
[2020-04-29T13:30:02.121Z] Caused: java.lang.InterruptedException: java.lang.InterruptedException: no matches found within 10000
[2020-04-29T13:30:02.121Z] 	at hudson.FilePath.act(FilePath.java:1071)
[2020-04-29T13:30:02.121Z] 	at hudson.FilePath.act(FilePath.java:1058)
[2020-04-29T13:30:02.121Z] 	at hudson.FilePath.validateAntFileMask(FilePath.java:2684)
[2020-04-29T13:30:02.121Z] 	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:265)
[2020-04-29T13:30:02.121Z] 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
[2020-04-29T13:30:02.121Z] 	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
[2020-04-29T13:30:02.121Z] 	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2020-04-29T13:30:02.121Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2020-04-29T13:30:02.121Z] 	at java.lang.Thread.run(Thread.java:748)
[2020-04-29T13:30:02.121Z] No artifacts found that match the file pattern "**/build/TEST*.out". Configuration error?
[2020-04-29T13:30:02.819Z] + curl -sSLo codecov https://codecov.io/bash
[2020-04-29T13:30:02.819Z] + FILE=auditbeat/build/coverage/full.cov
[2020-04-29T13:30:02.819Z] + [ -f auditbeat/build/coverage/full.cov ]
[2020-04-29T13:30:02.819Z] + FILE=filebeat/build/coverage/full.cov
[2020-04-29T13:30:02.819Z] + [ -f filebeat/build/coverage/full.cov ]
[2020-04-29T13:30:02.819Z] + bash codecov -f filebeat/build/coverage/full.cov
[2020-04-29T13:30:02.819Z] 
[2020-04-29T13:30:02.819Z]   _____          _
[2020-04-29T13:30:02.819Z]  / ____|        | |
[2020-04-29T13:30:02.819Z] | |     ___   __| | ___  ___ _____   __
[2020-04-29T13:30:02.819Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-04-29T13:30:02.819Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-04-29T13:30:02.819Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-04-29T13:30:02.819Z]                               Bash-tbd
[2020-04-29T13:30:02.819Z] 
[2020-04-29T13:30:02.819Z] 
[2020-04-29T13:30:02.819Z] ==> Jenkins CI detected.
[2020-04-29T13:30:02.819Z]     project root: .
[2020-04-29T13:30:02.819Z]     Fixing merge commit SHA
[2020-04-29T13:30:03.164Z]     Yaml found at: codecov.yml
[2020-04-29T13:30:03.165Z]     -> Found 1 reports
[2020-04-29T13:30:03.165Z] ==> Detecting git/mercurial file structure
[2020-04-29T13:30:03.165Z] ==> Reading reports
[2020-04-29T13:30:03.165Z]     + filebeat/build/coverage/full.cov bytes=259428
[2020-04-29T13:30:03.165Z] ==> Appending adjustments
[2020-04-29T13:30:03.165Z]     http://docs.codecov.io/docs/fixing-reports
[2020-04-29T13:30:12.145Z]     + Found adjustments
[2020-04-29T13:30:12.145Z] ==> Gzipping contents
[2020-04-29T13:30:12.145Z] ==> Uploading reports
[2020-04-29T13:30:12.145Z]     url: https://codecov.io
[2020-04-29T13:30:12.145Z]     query: branch=PR-17890&commit=d272f7f160991951c9cb06dcbcba80596c5aed36&build=6&build_url=https%3A%2F%2Fbeats-ci.elastic.co%2Fjob%2FBeats%2Fjob%2Fbeats-beats-mbp%2Fjob%2FPR-17890%2F6%2F&name=&tag=&slug=elastic%2Fbeats&service=jenkins&flags=&pr=17890&job=
[2020-04-29T13:30:12.146Z]     -> Pinging Codecov
[2020-04-29T13:30:12.146Z] https://codecov.io/upload/v4?package=bash-tbd&token=secret&branch=PR-17890&commit=d272f7f160991951c9cb06dcbcba80596c5aed36&build=6&build_url=https%3A%2F%2Fbeats-ci.elastic.co%2Fjob%2FBeats%2Fjob%2Fbeats-beats-mbp%2Fjob%2FPR-17890%2F6%2F&name=&tag=&slug=elastic%2Fbeats&service=jenkins&flags=&pr=17890&job=
[2020-04-29T13:30:12.146Z] HTTP 400
[2020-04-29T13:30:12.146Z] Please provide the repository token to upload reports via `-t :repository-token`
[2020-04-29T13:30:12.146Z] + FILE=heartbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f heartbeat/build/coverage/full.cov ]
[2020-04-29T13:30:12.146Z] + FILE=libbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f libbeat/build/coverage/full.cov ]
[2020-04-29T13:30:12.146Z] + FILE=metricbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f metricbeat/build/coverage/full.cov ]
[2020-04-29T13:30:12.146Z] + FILE=packetbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f packetbeat/build/coverage/full.cov ]
[2020-04-29T13:30:12.146Z] + FILE=winlogbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f winlogbeat/build/coverage/full.cov ]
[2020-04-29T13:30:12.146Z] + FILE=journalbeat/build/coverage/full.cov
[2020-04-29T13:30:12.146Z] + [ -f journalbeat/build/coverage/full.cov ]
[2020-04-29T13:30:14.853Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats-beats-mbp_PR-17890
[2020-04-29T13:30:15.557Z] [INFO] getVaultSecret: Getting secrets
[2020-04-29T13:30:15.745Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-04-29T13:30:16.968Z] + chmod 755 generate-build-data.sh
[2020-04-29T13:30:16.969Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17890/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17890/runs/6 FAILURE 3920583
[2020-04-29T13:30:17.519Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17890/runs/6/steps/?limit=10000 -o steps-info.json
[2020-04-29T13:30:18.863Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats-beats-mbp/PR-17890/runs/6/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--PIPELINE
{"commit":"abf6f9a036b915cd3a5c3a31ccbf59f51058710c","number":"6","status":"FAILURE","url":"https://beats-ci.elastic.co/job/Beats/job/beats-beats-mbp/job/PR-17890/6/"}
PIPELINE-->

leehinman commented on x-pack/filebeat/module/fortinet/firewall/ingest/traffic.json:
```diff
@@ -0,0 +1,506 @@
+    {
+      "append": {
+        "field": "event.type",
+        "value": "network",
```
network isn't an event type, it is an event category though

P1llus replied:
Typo, should have been protocol

leehinman commented on x-pack/filebeat/module/fortinet/firewall/ingest/traffic.json:
```diff
@@ -0,0 +1,506 @@
+      "append": {
+        "field": "event.type",
+        "value": "error",
+        "if": "['dns', 'ip-conn'].contains(ctx.fortinet?.firewall?.action)"
```
this is interesting, it might be worth switching format to yaml so we can easily have a comment that says why dns & ip-conn actions are errors.

P1llus replied:
Next commit will come in YAML, added a JSON commit first so its easier to see the changes I did between the newest JSON and the earlier. The mapping however seems to be correct.

P1llus replied:
This seems to have been a typo from me.

leehinman commented on x-pack/filebeat/module/fortinet/firewall/ingest/event.json:
```diff
@@ -0,0 +1,518 @@
+    {
+      "rename": {
+        "field": "fortinet.firewall.hostname",
+        "target_field": "url.domain",
```
Another interesting mapping that could really use a comment

P1llus replied:
Next commit will come in YAML, added a JSON commit first so its easier to see the changes I did between the newest JSON and the earlier

leehinman commented on x-pack/filebeat/module/fortinet/firewall/ingest/event.json:
```diff
@@ -0,0 +1,518 @@
+          "allowed",
+          "start"
+        ],
+        "if": "['FSSO-logon', 'auth-logon'].contains(ctx.fortinet?.firewall?.action)"
```
I think you want category authentication, type of start & outcome success for a logon event, but maybe this represents more than that?

P1llus replied:
This has been modified. The category is handled by checking for the "user" subtype a bit longer down the pipelien

leehinman commented on x-pack/filebeat/module/fortinet/firewall/ingest/event.json:
```diff
@@ -0,0 +1,518 @@
+          "allowed",
+          "end"
+        ],
+        "if": "['FSSO-logoff', 'auth-logout'].contains(ctx.fortinet?.firewall?.action)"
```
I think you want category authentication, type of end, & outcome success for a logout event, but maybe this represents more than that?

P1llus replied:
This has been modified. The category is handled by checking for the "user" subtype a bit longer down the pipeline

leehinman commented on x-pack/filebeat/module/fortinet/firewall/test/fortinet.log-expected.json:
```diff
@@ -0,0 +1,1488 @@
+        "tags": [
+            "fortinet-firewall"
+        ],
+        "url.domain": "elastichostname",
```
this doesn't seem right.  shouldn't the url.domain be "example.co" for this?

webmat commented on x-pack/filebeat/module/fortinet/_meta/fields.yml:
```diff
@@ -0,0 +1,10 @@
+- key: fortinet
+  title: Fortinet
+  description: >
+    fortinet Module
+  fields:
+    - name: fortinet
+      type: group
+      description: >
+        Fields from fortinet FortiOS
+      fields:
```
I think you can insert the crc32 field like this. Make sure you prepend. You'll notice the yaml file ends with a dangling `fields:` key, and is missing the value for that key. It's because each dataset's yaml file is appended there.

```suggestion
- key: file
  title: File
  fields:
    - name: hash.crc32
      type: keyword
      description: >
        The most modern hashing algorithm in history
        
- key: fortinet
  title: Fortinet
  description: >
    fortinet Module
  fields:
    - name: fortinet
      type: group
      description: >
        Fields from fortinet FortiOS
      fields:
```

andrewkroh commented on x-pack/filebeat/module/fortinet/_meta/docs.asciidoc:
```diff
@@ -7,13 +7,17 @@
 
 This is a module for Fortinet FortiOS logs sent in the syslog format.
 
+To configure a remote syslog destination, please reference the https://docs.fortinet.com/document/fortigate/6.0.0/cli-reference/260508/log-syslogd-syslogd2-syslogd3-syslogd4-setting[Fortigate/FortiOS Documentation]
```
```suggestion
To configure a remote syslog destination, please reference the https://docs.fortinet.com/document/fortigate/6.0.0/cli-reference/260508/log-syslogd-syslogd2-syslogd3-syslogd4-setting[Fortigate/FortiOS Documentation].
```

andrewkroh commented on x-pack/filebeat/module/fortinet/_meta/docs.asciidoc:
```diff
@@ -7,13 +7,17 @@
 
-This module has been tested against FortiOS version 6.0 and above
+This module has been tested against FortiOS version 6.0.x and 6.2.x. 
+Versions above this is expected to work but not tested.
```
```suggestion
Versions above this are expected to work but have not been tested.
```

P1llus commented:
Okay phew, going to try to summarise the latest changes.

Could someone help me add a custom ECS field for "file.hash.crc32" ? I have tried changing the fields.yml to several different versions but it always breaks my nosetests no matter what.

Added the documentation changes from @andrewkroh 
Added lots of small tweaks to ECS to get even better mapping, should cover almost all types of events now, including outcome, kind, category and type.
Added changes from @leehinman , comments where appropriate.
Converted pipelines to yaml.
Reran nosetests and generated new expected output.
Tested the filebeat against a private dataset working just fine.

P1llus commented:
jenkins, test please

P1llus commented:
jenkins, test this please

adriansr commented on CHANGELOG.next.asciidoc:
```diff
@@ -281,6 +281,8 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Improve ECS categorization field mappings in misp module. {issue}16026[16026] {pull}17344[17344]
 - Added Unix stream socket support as an input source and a syslog input source. {pull}17492[17492]
+- Added new Fortigate Syslog filebeat module. {pull}17890[17890]
+- Improve ECS categorization field mappings for nginx module. {issue}16174[16174] {pull}17844[17844]
```
Is this non-related entry on purpose?

adriansr commented on x-pack/filebeat/module/fortinet/firewall/manifest.yml:
```diff
@@ -0,0 +1,21 @@
+  - name: tags
+    default: [fortinet-firewall]
+  - name: syslog_port
+    default: 9001
```
This default already collides with other 4 filesets. Can you change it to `9004` which is unused?

adriansr commented on x-pack/filebeat/module/fortinet/firewall/manifest.yml:
```diff
@@ -0,0 +1,21 @@
+    default: 9001
+  - name: input
+    default: syslog
+  - name: log_level
```
Seems to me that this variable is not used, just carried over from using Cisco's module as a base.

adriansr commented on x-pack/filebeat/modules.d/fortinet.yml.disabled:
```diff
@@ -0,0 +1,21 @@
+    # Set the log level from 1 (alerts only) to 7 (include all messages).
+    # Messages with a log level higher than the specified will be dropped.
+    # See https://www.cisco.com/c/en/us/td/docs/security/asa/syslog/b_syslog/syslogs-sev-level.html
+    #var.log_level: 7
```
Also this is unused and references Cisco ASA docs.

P1llus commented:
@adriansr I have added your changes, if you could take a look if they match? For some reason nosetests does not work anymore after pulling in the new master, I did test them also without any modifications and they still fail for some reason. Did work on all earlier commits as I always run it before a commit.

adriansr commented:
The tests are working for me 👍

elasticmachine commented:
## :green_heart: Build Succeeded 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/14//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/14//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/14//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-17890/detail/PR-17890/14//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_17890.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #17890 updated]
* Start Time: 2020-05-04T16:37:47.860+0000

* Duration: 90 min 33 sec (5373121)


* Commit: 529b72b7051cddfe6b91fd93eee2aafb69a946f3



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2769  |
| Skipped      | 404 |
| Total        | 3173   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->

P1llus commented:
> The tests are working for me

Got them to run now, forgot to add MODULE_PATH env. Everything else looks okay @adriansr ?

If you look at the changes to CHANGELOG.next.asciidoc, I do believe that it is the wrong approach correct?

leehinman commented on x-pack/filebeat/module/fortinet/firewall/test/fortinet.log-expected.json:
```diff
@@ -0,0 +1,1524 @@
+            "fortinet-firewall"
+        ],
+        "url.domain": "elastic.co",
+        "url.path": "https://example.com/"
```
What is currently in url.path I would expect to be in url.full or url.original since it looks like a full url.  For url.path I would expect "/" with a url.full of "https://example.com/".  Also it still looks like the url.domain and what is the full url don't match.  For a url.full of "https://example.com/" I would expect url.domain to be "example.com"

P1llus replied:
This was a typo in my dataset and not the mapping, fixing it now

nicpenning commented:
I will piggy back on @enotspe comment above.

Please add the logic to get url.full when the log type is utm and sub type is virus, instead of using url.path.

Let us know if we need to clarify.

whataboutpereira commented:
Hello! DNS requests can include multiple IP addresses in the form of

```ipaddr="192.168.200.11, 192.168.200.12"```

Fabricated log line for testing:

```<189>date=2020-04-23 time=12:17:04 devname="testswitch1" devid="somerouterid" logid="1501054802" type="utm" subtype="dns" eventtype="dns-response" level="notice" vd="root" eventtime=1587230224712900694 tz="-0500" policyid=26 sessionid=5432 srcip=192.168.2.1 srcport=54438 srcintf="port1" srcintfrole="lan" dstip=8.8.8.8 dstport=53 dstintf="wan1" dstintfrole="wan" proto=17 profile="elastictest" xid=2352 qname="elastic.co" qtype="A" qtypeval=1 qclass="IN" ipaddr="8.8.8.8, 8.8.4.4" msg="Domain is monitored" action="pass" cat=93 catdesc="Remote Access"```

sidahmed-malaoui commented:
@P1llus You said other products from fortinet will be available later on. Any idea when they will be available please ? I would really appreciate to get this information, even if it's an approximate date.

Thank you in advance for your response Sir.

P1llus commented:
@sidahmed-malaoui Unfortunately there is no current way to say this. Though I have created issues for some of them, and you are always free to follow those issues to get any information on updates. If there is a specific product missing, you can create your own issue and request it and we can see if there is any possibilities to add it in.

As always there is no real promise or timeline on this, it was rather an explanation that if we are to support multiple fortinet products then it would be in a separate fileset.

Examples: 
https://github.com/elastic/beats/issues/19314
https://github.com/elastic/beats/issues/19315

There is also quite some documentation around fortisandbox and fortiweb, but i'm still researching if those will be possible

enotspe commented:
> @P1llus You said other products from fortinet will be available later on. Any idea when they will be available please ? I would really appreciate to get this information, even if it's an approximate date.
> 
> Thank you in advance for your response Sir.

There are some logstash pipelines for [fortimail and fortiweb](https://github.com/enotspe/fortinet-2-elasticsearch) if that helps
